### DISCUSSION TITLE
Unable to detect objects using Yolov5

### DISCUSSION THREAD

**Discussion by ThenoobMario** [AxisCommunications/acap-computer-vision-sdk-examples#175] [Q&A] [answered]
After following the [guide](https://github.com/AxisCommunications/axis-model-zoo/blob/main/docs/yolov5-on-artpec8.md) in axis-model-zoo, I was able to load the model and run it on our P3265 with 11.8 Firmware. However, I am getting `0 objects found` in the output using the CV-sdk's object-detector-python example (With modifications in the `detect` method).

Considering the format of yolov5 output `[xc, yc, w, h, conf, class_1_conf, class_2_conf, ...., class_n_conf]`, we parsed the output and got the following result.
![image](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/assets/60312074/8f33b838-f110-43bb-99b2-484bc5f8c3af)

Where the array represents the first 5 values of the `conf` index.  

I need help in decoding the output of the yolov5 tflite model.

**Comment by Corallo** [comment]
Hi,

The first 5 "entries" of your output may have zero likelihood.
Try to sort the results by "conf" and display top n entries.

To debug your model, I would suggest to test first the output of the tflite on your local machine.

**Reply by ThenoobMario** [reply]
Hi @Corallo, 

Thanks for the reply. We tested the model on a local machine and we received the output as intended. I will sort the results by `conf` and see what can be the issue.

**Reply by ThenoobMario** [reply, marked as answer]
Hi @Corallo,

Thank you for pointing out the sorting of 'conf'. As it turns out, we were not processing the output tensor correctly and getting no results.

**Comment by AlejandroDiazD** [comment]
Hi @Corallo ,

I would like to ask you about working with Yolo models on that cameras as I'm currently a little bit stoked on an issue.

I'm trying to run inference on a LVE-P3268 camera with yolov5n.tflite exported model. The output I receive from processing an image has a shape of [1, 25200, 85]. This is a sample of the output:
[[[ 2 3 7 ... 0 0 0]
[ 2 3 7 ... 0 0 0]
[ 2 3 7 ... 0 0 0]
...
[230 232 19 ... 2 0 0]
[228 233 26 ... 2 0 0]
[228 232 47 ... 2 0 1]]]

About the dimensions, I already understood that:
* 1: is due to that only 1 image has been processed
* 25200: is the number of anchors
* 85: is [x, y, w, h, objectness, 80 coco classes]

The point is that I'm not being able to process that output to obtain human readable results. Do I have to apply any specific post processing? As the output seems to be quantized. I've tried to apply sigmoid to objectness and classes confidence, but the results of the objectness doesn't make sense as most of the values are almost 1.

Thank you in advance and best regards!

**Reply by Corallo** [reply]
Hi @AlejandroDiazD 

Take a look at this post here:
https://github.com/AxisCommunications/axis-model-zoo/discussions/45
I can't guarantee the code works out of the box, but it should offer some degree of guidance on how to do post-processing.
It is very new, so feedback is welcome!

**Reply by AlejandroDiazD** [reply]
Thank you @Corallo , this is exactly what I was talking about. I will continue the continue the conversation in the other post if it is fine for you.

**Reply by Corallo** [reply]
Some extra comment,

- Not sure if this applies to you, but I will say it just in case, the coco dataset is a benchmark dataset, it is probably not a good idea to train only on that. Camera frames look quite differently from the type of images you find in the coco dataset. If you used only that data as training, maybe that is a reason for the performance issue you get. In that case, you can try to evaluate the output of the model on one of the validation images of the coco dataset.
- It doesn't make sense to apply the sigmoid to all the "objectness". It could make sense to do it on the classes, but one detection at the time, but if you just want to know what class is that object, you can just pick the one with higher value, without bothering with the softmax.
- The fact that the "objectness" is very low for most of the detection is not a problem, it is actually normal, only a few detections should be higher, according to the presence of objects in the scene.
